crates/valence/src/weather.rs

```diff
@@ -0,0 +1,337 @@
+            kind: GameEventKind::BeginRaining,
+            value: f32::default(),
+        });
+    }
+
+    /// Sends the end rain event to the client.
+    fn end_raining(&mut self) {
+        self.write_packet(&GameStateChangeS2c {
+            kind: GameEventKind::EndRaining,
+            value: f32::default(),
+        });
+    }
+
+    /// Sends the set rain level event to the client.
+    fn set_rain_level(&mut self, level: f32) {
+        self.write_packet(&GameStateChangeS2c {
+            kind: GameEventKind::RainLevelChange,
+            value: level.clamp(WEATHER_LEVEL.start, WEATHER_LEVEL.end),
+        });
+    }
+
+    /// Sends the set thunder level event to the client.
+    fn set_thunder_level(&mut self, level: f32) {
+        self.write_packet(&GameStateChangeS2c {
+            kind: GameEventKind::ThunderLevelChange,
+            value: level.clamp(WEATHER_LEVEL.start, WEATHER_LEVEL.end),
+        });
+    }
+
+    /// Sends weather level events to the client.
+    fn set_weather(&mut self, weather: &Weather) {
+        if let Some(rain_level) = weather.rain {
+            self.set_rain_level(rain_level)
+        }
+
+        if let Some(thunder_level) = weather.thunder {
+            self.set_thunder_level(thunder_level)
+        }
+    }
+}
+
+fn handle_weather_begin_per_instance(mut query: Query<&mut Instance, Added<Weather>>) {
+    query.for_each_mut(|mut instance| {
+        instance.begin_raining();
+    });
+}
+
+fn handle_weather_end_per_instance(
+    mut query: Query<&mut Instance>,
+    mut removed: RemovedComponents<Weather>,
+) {
+    removed.iter().for_each(|entity| {
+        if let Ok(mut instance) = query.get_mut(entity) {
+            instance.end_raining();
+        }
+    })
+}
+
+fn handle_weather_change_per_instance(
+    mut query: Query<(&mut Instance, &Weather), Changed<Weather>>,
+) {
+    query.for_each_mut(|(mut instance, weather)| {
+        instance.set_weather(weather);
+    });
+}
+
+fn handle_weather_for_joined_player(
+    mut clients: Query<(&mut Client, &McEntity), Added<Client>>,
+    weathers: Query<&Weather, With<Instance>>,
+) {
+    clients.for_each_mut(|(mut client, mc_ent)| {
+        if let Ok(weather) = weathers.get(mc_ent.instance()) {
+            client.set_weather(weather);
+        }
+    })
+}
+
+fn handle_weather_begin_per_client(mut query: Query<&mut Client, Added<Weather>>) {
+    query.for_each_mut(|mut client| {
+        client.begin_raining();
+    });
+}
+
+fn handle_weather_end_per_client(
+    mut query: Query<&mut Client>,
+    mut removed: RemovedComponents<Weather>,
+) {
+    removed.iter().for_each(|entity| {
+        if let Ok(mut client) = query.get_mut(entity) {
+            client.end_raining();
+        }
+    })
+}
+
+fn handle_weather_change_per_client(mut query: Query<(&mut Client, &Weather), Changed<Weather>>) {
+    query.for_each_mut(|(mut client, weather)| {
+        client.set_weather(weather);
+    });
+}
```

Comment:
What happens if an entity has both client components and an `Instance` component? Changes to `Weather` will be sent twice -- once for the client and once for the instance. Looks like you'll want to add `Without<Instance>` to these systems.